### DISCUSSION TITLE
can someone please tell if its better than zsh and if yes how if no then why should we use nushell

### DISCUSSION THREAD

**Discussion by Greatz08** [nushell/nushell#13723] [General]
Asking here because i wanted to know from practical experience of others because i did tried zsh and did setup plugins and other cool stuff and its working very well for me tbh for almost an year so today i came across this cool project so thought to get experience of others on how it is and how is it different or better than zsh and hopefully without any biasness i can get better response :-))
Thankyou very much in advance

**Comment by Lockszmith-GH** [comment]
This is completely different.
Where zsh is "like bash, but a bit different", this one goes in a completley different direction.

Some of the things zsh has plugins for, will be built into the shell. Some of the things will be impossible or if possible, nobody ever wrote an implmentation for it yet. There are libraries out there that extend nushell, but it's not as vibrant as the zsh world - at least not yet.

I've been playing around with nushell for a while, still havn't fully committed.
I've been with bash for a very long time, even longer with PowerShell (on Windows), I'm currently switching between zsh and nushell on my linux machines.
nushell is a lot more like PowerShell than bash - but it's also very different. The syntax and assumptions it makes might not be intuitive to you at first.

With all that stated. Would I recommend this? YES, but prepare for a bumpy ride

Spin it up and experiment with it, and see if you like it - it will not have the same feel.
You definintly will not be able to copy paste shell code and have it 'just work', you'll need to adjust it or run with bash/sh when that happens.

Understand that nushell hasn't reached v1 yet, and it probably won't be anytime soon.
There are still aspects that don't function or don't function the way you'd expect them - or they work, just require cumbersome syntax.

**Reply by Greatz08** [reply]
Thankyou @Lockszmith-GH  for this detailed reply based on your experience. I will test and see but i do know that i will become back to zsh again :-))

**Comment by ghost** [comment]
Hi,

there most definitely is no straight answer to "which one is better in all cases".  
while both are shells nushell has a very distinct and different style compared to POSIX shells (bash, zsh, etc).  
its personal preference and usecase.  
a few comparison points:

* nushell can easily on the fly work with json, sqlite, etc.
* anything that requires loops is way easier in nushell (example: go into each directory and git push)
* tab-completion, etc is usually better in zsh (yes you can complete most things in nu, but it requires more configuration, etc). when i have to work with `systemctl`, etc i usually quickly switch to bash and then go back
* zsh is more reliable (yes nu is written in rust and fairly stable, but nu crashes on me from time to time)
* filtering, etc is a lot easier in nu (try to delete all files greater than 1GB with zsh without googling - sounds uncommon, but if you start thinking in nu a lot of things get easier)
* nu is often lot more verbose (aka longer) (`ls > file_list.txt` -> `^ls | save -f file_list.txt`)
* fallback commands, error-handling, etc are easier (or at least more compact) in zsh

for me learning nu was definitely worth it. the web-api exploration, sqlite manipulation/reading, csv to json, etc on its own makes it worth it.

therefore i use them as follows:

* on my programing pc i have nu as default since i often work with web-apis, json, etc.
  * most common complex commands are in nu-history and therefore `nmcli c up MyWifi` completes when i type `nm`.
  * when i use something less common with long texts i switch to bash.
* on my server i have bash as default since i often restart services, use docker, etc and make heavy use of tab completion.
  * when i mass-move/change files, work with json, etc i just start nu.
* for scripts i use POSIX (sh, not bash) if its simple since this increases portability. if i don't intend to move it anywhere else or its more complex i use nu since it has a lot less footguns (see shellcheck) and most my scripts work with json / structured data.

**Reply by Greatz08** [reply]
Thankyou @Jan9103  for sharing your experience and the way you use things for different usecase
